README.md
## Repository Overview

- This project is a development of a small set of [Backbase Service SDK](https://community.backbase.com/documentation/ServiceSDK/latest/index) (**Spring Boot** and **Cloud**) based Microservices projects that implement cloud-native intuitive, Microservices design patterns, and coding best practices.
- The project follows [**CloudNative**](https://www.cncf.io/) recommendations and the [**twelve-factor app**](https://12factor.net/) methodology for building *software-as-a-service apps* to show how μServices should be developed and deployed.
- This project uses technologies broadly used in Backbase.Like Docker, Kubernetes, Java SE 11, Spring Boot, Spring Cloud
- 'cards-presentation-service' has been generated using `core-service-archetype` - [Community guide](https://community.backbase.com/documentation/ServiceSDK/latest/create_a_core_service)
- This service implements logic for the UI components (Widget/Journey) to view and manage bank customer cards. Service integrates with [Marqeta](https://www.marqeta.com) card issuer.
- The service is implementation of `card-manager-client-api` - https://repo.backbase.com/specs/card-manager/
- Refer to [workflow guide](../../../docs/tree/master/backend) for Backend CI Workflow documentation

---
## Repository Description
### Project Structure
The project structure for each custom integration service follows the pattern as described below :

```
.
├── .github                       # All GitHub Actions files
│   ├── ISSUE_TEMPLATE            # Templates for 'major','minor','patch' releases
│   └── workflows                 # GitHub Actions workflows for CI
├── src                           # Source and Unit Test files
    ├── main                      # Application container projects
    │   ├── java/com/backbase/productled
    │   │   ├── api               # Controller classes
    │   │   │   └── ...
    │   │   ├── config            # Configuration classes
    │   │   │   └── ...
    │   │   ├── mapper            # Model classes
    │   │   │   └── ...
    │   │   └── service           # Service classes
    │   │       └── ...
    │   └── resources             # All resource files except core classes
    │       └── ...
    └── test                      # JUnit test file
        └── ...
```

To view individual classes for this repository, select relevant branch from the GitHub UI and then press ‘.'
This will open the GitHub Web Editor.Alternatively, you can also access the Web Editor by changing .com to .dev in the URL.

Expand each file in the Web Editor for explanation and purpose.

## Repository Configurations
### DSC (basic-installation.yml) configuration

```yaml
custom:
  enabled: true
  services:
    cards-presentation-service:
      enabled: true
      app:
        image:
          tag: "1.0.1-SNAPSHOT"
          repository: cards-presentation-service
      database: false
      livenessProbe:
        enabled: true
      readinessProbe:
        enabled: true
      env:
        BACKBASE_COMMUNICATION_HTTP_ACCESS-TOKEN-URI: "http://token-converter:8080/oauth/token"
        marqeta.username:
          valueFrom:
            secretKeyRef:
              name: marqeta-credentials
              key: username
        marqeta.password:
          valueFrom:
            secretKeyRef:
              name: marqeta-credentials
              key: password
        marqeta.baseUrl:
          valueFrom:
            configMapKeyRef:
              name: marqeta-env
              key: basePath
```
#### Marqeta config
The following properties **must** be set as they have no default:

Property | Description
--- | ---
**marqueta.baseUrl** | Base URL of Marqueta API (i.e. `https://sandbox-api.marqeta.com/v3`)
**marqueta.username** | API username
**marqueta.password** | API password

#### Stream config (if applicable)
Configuration properties used by Stream to interact with Backbase services (DBS, Token converter...).

Not applicable for 'cards-presentation-service'
---
## Customisation in project
CVV validation is not done as part this project due to testing complexity but can be added using marqeta getCvv api if required.
### Configuration changes
If project adopts Marqeta, only Marqeta configuration properties has to be changed.

### Component changes
If project adopts a different vendor than Marqeta, all services from `com.backbase.productled.service` package have to be replaced.

---
## Getting Started
### BaaS setup

- [ ] Step 1: Modify https://github.com/baas-devops-reference/ref-self-service/blob/main/self-service.tfvars by adding to `ecr` list name of new repository: `cards-presentation-service'
- [ ] Step 2: Checkout the following repository: https://github.com/baas-devops-reference/ref-applications-live/blob/main/runtimes/dev/basic-installation.yaml apply your deployment configurations example see _DSC (basic-installation.yml) configuration_ above.
- [ ] Step 3: Run the pre-commit to validate the configurations => ` pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Step 4: Commit and Push your changes; wait for the template rendering and lint jobs to complete
- [ ] Step 5: Merge into `master` to trigger deployment.

### Local setup

- [ ] Step 1: Ensure to check the prerequisites for [local developer environment](https://community.backbase.com/documentation/ServiceSDK/latest/create_developer_environment)
- [ ] Step 2: Run command => `mvn spring-boot:run -Dspring.profiles.active=local`
- [ ] Step 3: To run the service from the built binaries, use => `java -jar target/cards-presentation-service-1.0.1-SNAPSHOT.jar -Dspring.profiles.active=local`

## Contributions
Please create a branch and a PR with your contributions. Commit messages should follow [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages)


CHANGELOG.md
# Changelog

## [Unreleased]

## [0.22] - 2022-07-22

### What has Changed

-   support for both ACR/ECR
-   after release build version has been reset to 0

## [0.21] - 2022-07-14

### What has Changed

-   upgrade backbase bom to 2022.07.1

## [0.20] - 2022-07-13

### What has Changed

-   upgrade backbase bom to 2022.07
-   versioning changes

## [0.19] - 2022-06-20

### What has Changed

-   Upgrade BOM to 2022.05.6

## [0.18] - 2022-05-13

### What has Changed

-   Upgrade BOM to 2022.04

## [0.17.1] - 2022-05-06

### Bug Fixes

-   Fixed pr workflow if condition.

## [0.17] - 2022-05-03

### What has Changed

-   Changes to Github Workflow

## [0.16] - 2022-03-02

### Bug Fixes

-   PROMO-3317 : Version bug fix

## [0.15] - 2022-02-25

### New/Improved

-   Added new workflow

## [0.14] - 2022-02-11

### What has Changed

-   Upgraded to 2022.01-LTS

## [0.13] - 2022-01-18

### What has Changed

-   Upgraded to 2021.12

## [0.12] - 2021-12-06

### What has Changed

-   Upgraded to 2021.10

## [0.11] - 2021-11-19

### New/Improved

-   Cards service-api implementation

## [0.10] - 2021-10-14

### What has Changed

-   Upgraded to 2021.09

## [0.9] - 2021-09-09

### New/Improved

-   Implemented BB bom changes and ssdk version upgrade
-   Moved project structure info from README to Javadocs / comments
-   Workflow improvement

### What has Changed

-   BB bom addition in dependency management to retrieve latest product/spec version

[Unreleased]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.22...HEAD

[0.22]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.21...0.22

[0.21]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.20...0.21

[0.20]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.19...0.20

[0.19]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.18...0.19

[0.18]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.17.1...0.18

[0.17.1]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.17...0.17.1

[0.17]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.16...0.17

[0.16]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.15...0.16

[0.15]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.14...0.15

[0.14]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.13...0.14

[0.13]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.12...0.13

[0.12]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.11...0.12

[0.11]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.10...0.11

[0.10]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.9...0.10

[0.9]: https://github.com/baas-devops-reference/cards-presentation-service/compare/5154031768bc4e9b0337569b0079a4265f81480e...0.9

## [0.8] - 2021-08-23

### New/Improved

-   Changes to return last cancelled card in Get Cards api response

### What has Changed

-   Replace card api to return cancelled card in response instead of new issues card 
-   Delivery status of new card is mocked with Delivered status
-   Get Card api will return 1 last cancelled card and other Active/InActive cards in response

### Bug Fixes

-   Android | Replace Card | app crashes when clicking Done after completing card replacement flow

[Unreleased]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.8...HEAD

[0.8]: https://github.com/baas-devops-reference/cards-presentation-service/compare/29ceaaeebac907ccedf70b5d6fbd2f25d221a4ba...0.8

## [0.7] - 2021-07-15

### No relevant changes

[Unreleased]: https://github.com/baas-devops-reference/cards-presentation-service/compare/0.7...HEAD

[0.7]: https://github.com/baas-devops-reference/cards-presentation-service/compare/4789badcc3c2d5f32bcdf0d3a3f9ed4e00ab8f55...0.7


src/main/java/com/backbase/productled/mapper/CardsMappers.java
package com.backbase.productled.mapper;

import static com.backbase.productled.mapper.MapperConstants.ACTIVE;
import static com.backbase.productled.mapper.MapperConstants.ATM;
import static com.backbase.productled.mapper.MapperConstants.CANCELLED;
import static com.backbase.productled.mapper.MapperConstants.CARD_HOLDER_NAME;
import static com.backbase.productled.mapper.MapperConstants.DELIVERED;
import static com.backbase.productled.mapper.MapperConstants.INACTIVE;
import static com.backbase.productled.mapper.MapperConstants.IN_TRANSIT;
import static com.backbase.productled.mapper.MapperConstants.NOT_UNDER_REPLACEMENT;
import static com.backbase.productled.mapper.MapperConstants.ONLINE;
import static com.backbase.productled.mapper.MapperConstants.ORDERED;
import static com.backbase.productled.mapper.MapperConstants.PROCESSED;
import static com.backbase.productled.mapper.MapperConstants.REPLACEMENT_REASON;
import static com.backbase.productled.mapper.MapperConstants.REPLACEMENT_STATUS;
import static java.util.Objects.requireNonNull;

import com.backbase.marqeta.clients.model.CardRequest;
import com.backbase.marqeta.clients.model.CardResponse;
import com.backbase.marqeta.clients.model.CardResponse.StateEnum;
import com.backbase.marqeta.clients.model.CardTransitionRequest;
import com.backbase.marqeta.clients.model.CardUpdateRequest;
import com.backbase.marqeta.clients.model.UserCardHolderResponse;
import com.backbase.marqeta.clients.model.UserCardHolderUpdateModel;
import com.backbase.marqeta.clients.model.VelocityControlListResponse;
import com.backbase.marqeta.clients.model.VelocityControlResponse;
import com.backbase.marqeta.clients.model.VelocityControlUpdateRequest;
import com.backbase.presentation.card.rest.spec.v2.cards.CardHolder;
import com.backbase.presentation.card.rest.spec.v2.cards.CardItem;
import com.backbase.presentation.card.rest.spec.v2.cards.Delivery;
import com.backbase.presentation.card.rest.spec.v2.cards.DeliveryTransitStep;
import com.backbase.presentation.card.rest.spec.v2.cards.DeliveryTransitStep.StatusEnum;
import com.backbase.presentation.card.rest.spec.v2.cards.LockStatus;
import com.backbase.presentation.card.rest.spec.v2.cards.Replacement;
import com.backbase.presentation.card.rest.spec.v2.cards.YearMonth;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;

/**
 * Maps models between Marqeta <-> DBS
 */
@Mapper(componentModel = "spring")
public interface CardsMappers {

    @Mapping(target = "id", source = "cardResponse.token")
    @Mapping(target = "maskedNumber", source = "cardResponse.lastFour")
    @Mapping(target = "brand", expression = "java(cardResponse.getMetadata().get(\"brand\"))")
    @Mapping(target = "type", expression = "java(cardResponse.getMetadata().get(\"type\"))")
    @Mapping(target = "subType", expression = "java(cardResponse.getMetadata().get(\"subType\"))")
    @Mapping(target = "holder", source = "cardResponse", qualifiedByName = "getCardHolder")
    @Mapping(target = "name", expression = "java(cardResponse.getMetadata().get(\"name\"))")
    @Mapping(target = "status", source = "cardResponse", qualifiedByName = "getStatus")
    @Mapping(target = "lockStatus", source = "cardResponse", qualifiedByName = "getLockStatus")
    @Mapping(target = "replacement", source = "cardResponse", qualifiedByName = "getReplacement")
    @Mapping(target = "expiryDate", source = "cardResponse", qualifiedByName = "getExpiryDate")
    @Mapping(target = "currency", expression = "java(cardResponse.getMetadata().get(\"currency\"))")
    @Mapping(target = "limits", expression = "java(cardLimits.getData().stream().map(velocityControlResponse -> {return new com.backbase.presentation.card.rest.spec.v2.cards.CardLimit().id(velocityControlResponse.getToken()).amount(velocityControlResponse.getAmountLimit().setScale(0, java.math.BigDecimal.ROUND_DOWN)).frequency(com.backbase.productled.utils.FrequencyEnum.fromValue(velocityControlResponse.getVelocityWindow().getValue()).getValue()).channel(getChannel(velocityControlResponse.getIncludeWithdrawals()).toLowerCase()).minAmount(java.math.BigDecimal.valueOf(Long.parseLong(cardResponse.getMetadata().get(getChannel(velocityControlResponse.getIncludeWithdrawals()).toLowerCase().concat(\"MinAmount\")))).setScale(0, java.math.BigDecimal.ROUND_DOWN)).maxAmount(java.math.BigDecimal.valueOf(Long.parseLong(cardResponse.getMetadata().get(getChannel(velocityControlResponse.getIncludeWithdrawals()).toLowerCase().concat(\"MaxAmount\")))).setScale(0, java.math.BigDecimal.ROUND_DOWN));}).collect(java.util.stream.Collectors.toList()))")
    @Mapping(target = "delivery", source = "cardResponse", qualifiedByName = "getDelivery")
    @Mapping(target = "additions", ignore = true)
    @Mapping(target = "ownership", ignore = true)
    CardItem mapCard(CardResponse cardResponse, VelocityControlListResponse cardLimits);

    default String getChannel(Boolean withdrawalsWindow) {
        return Boolean.TRUE.equals(withdrawalsWindow) ? ATM : ONLINE;
    }

    @Mapping(target = "token", expression = "java(java.util.UUID.randomUUID().toString())")
    @Mapping(target = "cardToken", source = "id")
    @Mapping(target = "channel", expression = "java(CardTransitionRequest.ChannelEnum.API)")
    @Mapping(target = "state", source = "state")
    @Mapping(target = "reason", ignore = true)
    @Mapping(target = "reasonCode", ignore = true)
    @Mapping(target = "validations", ignore = true)
    CardTransitionRequest mapCardTransitionRequest(String id,
        com.backbase.marqeta.clients.model.CardTransitionRequest.StateEnum state);

    @Mapping(target = "token", source = "token")
    @Mapping(target = "metadata", source = "token", qualifiedByName = "activationMetaData")
    @Mapping(target = "userToken", ignore = true)
    @Mapping(target = "fulfillment", ignore = true)
    CardUpdateRequest mapUpdateCardRequestForActivation(String token);

    UserCardHolderUpdateModel mapUpdateCardHolderRequest(UserCardHolderResponse response);

    default CardRequest mapCreateCardRequest(CardResponse cardDetails) {
        return new CardRequest()
            .cardProductToken(cardDetails.getCardProductToken())
            .userToken(cardDetails.getUserToken())
            .metadata(cardDetails.getMetadata());
    }

    @Named("activationMetaData")
    default Map<String, String> getActivationMetaData(String token) {
        HashMap<String, String> metaData = new HashMap<>();
        metaData.put(REPLACEMENT_STATUS, NOT_UNDER_REPLACEMENT);
        metaData.put(REPLACEMENT_REASON, null);
        return metaData;
    }

    @Named("getStatus")
    default String getStatus(CardResponse cardResponse) {
        if (Arrays.asList(StateEnum.ACTIVE, StateEnum.SUSPENDED).contains(cardResponse.getState())) {
            return ACTIVE;
        } else if(StateEnum.UNACTIVATED == cardResponse.getState()) {
            return INACTIVE;
        }
        return CANCELLED;
    }

    @Named("getLockStatus")
    default LockStatus getLockStatus(CardResponse cardResponse) {
        if (StateEnum.SUSPENDED == cardResponse.getState()) {
            return LockStatus.LOCKED;
        }
        return LockStatus.UNLOCKED;
    }

    @Named("getReplacement")
    default Replacement getReplacement(CardResponse cardResponse) {
        return new Replacement()
            .status(requireNonNull(cardResponse.getMetadata()).get(REPLACEMENT_STATUS))
            .reason(requireNonNull(cardResponse.getMetadata()).get(REPLACEMENT_REASON));
    }

    @Named("getExpiryDate")
    default YearMonth getExpiryDate(CardResponse cardResponse) {
        return new YearMonth()
            .year(String.valueOf(cardResponse.getExpirationTime().getYear()))
            .month(String.format("%02d", cardResponse.getExpirationTime().getMonthValue()));

    }

    @Named("getCardHolder")
    default CardHolder getCardHolder(CardResponse cardResponse) {
        return new CardHolder()
            .name(requireNonNull(cardResponse.getMetadata()).get(CARD_HOLDER_NAME));
    }

    @Named("getDelivery")
    default Delivery getDelivery(CardResponse cardResponse) {
        if (cardResponse.getState() == StateEnum.UNACTIVATED) {
            return new Delivery()
                .transitSteps(Arrays.asList(
                    new DeliveryTransitStep().name(ORDERED).status(StatusEnum.SUCCESS)
                        .stepDateTime(OffsetDateTime.now().minusMinutes(30)),
                    new DeliveryTransitStep().name(PROCESSED).status(StatusEnum.SUCCESS)
                        .stepDateTime(OffsetDateTime.now().minusMinutes(25)),
                    new DeliveryTransitStep().name(IN_TRANSIT).status(StatusEnum.SUCCESS)
                        .stepDateTime(OffsetDateTime.now().minusMinutes(15)),
                    new DeliveryTransitStep().name(DELIVERED).status(StatusEnum.SUCCESS)
                        .stepDateTime(OffsetDateTime.now().minusMinutes(5))));
        }
        return null;
    }

    @Mapping(target = "amountLimit.", source = "amount")
    VelocityControlUpdateRequest mapVelocityControlUpdateRequest(VelocityControlResponse velocityControlResponse,
        BigDecimal amount);

    com.backbase.presentation.card.api.service.v2.model.CardItem mapToServiceCardItem(CardItem cardItem);
}
